### PR TITLE
feat: add tests to wallet page + pool page

### LINKED ISSUE #256: add tests to pool add liquidity page

### LINKED ISSUE #259: Add tests to Wallet page

### PR DESCRIPTION
**Pull request by LuizAsFight**
Adding specific test to the pages
- welcome (WelcomeTest.test.tsx)
- pool page (PoolPage.test.tsx)

Added end-to-end test in file `App.test.tsx`. **Objective**: make sure that whole application is working from a brand-new created fuel-core node. For now we have the below tests, but will be nice to add more as validate the whole app in an ideal world:
- Create Wallet
- Add funds
- Go to Pool Page
- Validate Inputs
- TODO: create mint tests (then validate create liquidity step)

Obs: Create liquidity step is already validated in PoolPage.test.tsx, but we need mint tests to include it in the end-to-end test in file `App.test.tsx`

Closes #259 
Closes #256

### CODE REVIEW

File: packages/app/src/components/FaucetApp.tsx [hunk 1]
```diff
@@ -46,6 +46,7 @@ export function FaucetApp({ onSuccess }: FaucetAppProps) {
         className="mt-5"
         isLoading={faucet.mutation.isLoading}
         onPress={() => faucet.handleFaucet(captcha.value)}
+        data-testid="addFundsBtn"
```
Comment (luizstacio) [inline packages/app/src/components/FaucetApp.tsx:49]: https://testing-library.com/docs/queries/bytestid

> Using data-testid attributes do not resemble how your software is used and should be avoided if possible

I think @pedronauck had add `aria-` on the other fields it improves the accessibility and helps on tests

**Review by luizstacio — CHANGES_REQUESTED** [review]

File: packages/app/package.json [hunk 2]
```diff
@@ -45,6 +45,7 @@
     "bignumber.js": "^9.0.2",
     "classnames": "^2.3.1",
     "clipboard": "^2.0.11",
+    "cross-fetch": "^3.1.5",
```
[the remark this answers is not in the document]
Comment (LuizAsFight) [reply packages/app/package.json:48]: for regular fetchs to work in tests without needing to moke the requests

File: packages/app/src/components/FaucetApp.tsx (hunk 1 above)
Comment (LuizAsFight) [reply packages/app/src/components/FaucetApp.tsx:49]: what aria has to do with data-testid? I dont understand what exactly you want with this comment

you want me to add aria prop as well to improve accessibility ?

File: packages/app/src/hooks/useWelcomeSteps.tsx [hunk 3]
```diff
@@ -139,16 +139,20 @@ export const stepsSelectors = {
...
-  const [state, send, service] = useMachine<Machine>(
-    welcomeStepsMachine.withConfig({
-      actions: {
-        navigateTo: (context) => {
-          if (context.current.id > 2) return;
-          navigate?.(`/welcome/${context.current.path}`);
+
+  const machine = useMemo(
+    () =>
+      welcomeStepsMachine.withConfig({
```
Comment (pedronauck) [inline packages/app/src/hooks/useWelcomeSteps.tsx:145]: dude, you don't need `useMemo` inside this, it's already has all performance handles inside `useMachine`

File: packages/app/src/hooks/useFaucet.ts [hunk 4]
```diff
@@ -45,7 +46,7 @@ export function useFaucet(opts: UseFaucetOpts = {}) {
       onSuccess: () => {
         // Navigate to assets page to show new cons
         // https:// github.com/FuelLabs/swayswap-demo/issues/40
-        refreshBalances();
+        balances.refetch();
```
Comment (pedronauck) [inline packages/app/src/hooks/useFaucet.ts:49]: @luizstacio changed this couple pull requests earlier, why did you change it again?

File: packages/app/src/components/FaucetApp.tsx (hunk 1 above)
Comment (pedronauck) [reply packages/app/src/components/FaucetApp.tsx:49]: this section has a good explanation about this
https://testing-library.com/docs/queries/byrole/

File: packages/app/src/components/Header.tsx [hunk 5]
```diff
@@ -69,13 +69,15 @@ const Header = () => {
                 icon={MdSwapCalls}
                 onPress={() => navigate(Pages.swap)}
                 isActive={location.pathname === Pages.swap}
+                data-testid="swapMenuBtn"
```
Comment (pedronauck) [inline packages/app/src/components/Header.tsx:72]: check if can replace `data-testid` here

File: packages/app/src/components/Header.tsx [hunk 6]
```diff
@@ -69,13 +69,15 @@ const Header = () => {
...
               >
                 Swap
               </HeaderNav>
               <HeaderNav
                 icon={BiDollarCircle}
                 onPress={() => navigate(Pages.pool)}
                 isActive={location.pathname.includes(Pages.pool)}
+                data-testid="poolMenuBtn"
```
Comment (pedronauck) [inline packages/app/src/components/Header.tsx:80]: check if can replace `data-testid` here

File: packages/app/src/pages/PoolPage/AddLiquidity.tsx [hunk 7]
```diff
@@ -214,7 +214,7 @@ export default function AddLiquidity() {
             reservesFromToRatio={reservesFromToRatio || addLiquidityRatio || 1}
           />
           <Button
-            data-testid="submit"
+            data-testid="addLiquiditySubmitBtn"
```
Comment (pedronauck) [inline packages/app/src/pages/PoolPage/AddLiquidity.tsx:217]: check if can replace `data-testid` here

File: packages/app/src/pages/PoolPage/Pools.tsx [hunk 8]
```diff
@@ -28,6 +28,7 @@ const SubpagesNav = () => {
         variant="primary"
         onPress={() => navigate(`../${Pages["pool.addLiquidity"]}`)}
         className={cx(style.button)}
+        data-testid="headerAddLiquidityBtn"
```
Comment (pedronauck) [inline packages/app/src/pages/PoolPage/Pools.tsx:31]: check if can replace `data-testid` here

File: packages/app/src/pages/WelcomePage/CreateWallet.tsx [hunk 9]
```diff
@@ -28,7 +28,12 @@ export default function CreateWallet() {
         To get started you&apos;ll need a wallet, <br />
         click below to generate one.
       </p>
-      <Button variant="primary" size="lg" onPress={handleCreateWallet}>
+      <Button
+        variant="primary"
+        size="lg"
+        onPress={handleCreateWallet}
+        data-testid="createWalletBtn"
```
Comment (pedronauck) [inline packages/app/src/pages/WelcomePage/CreateWallet.tsx:35]: check if can replace `data-testid` here

File: packages/app/src/pages/WelcomePage/WalletPage.test.tsx [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+  fireEvent,
+  renderWithRouter,
+  screen,
+  waitFor,
+} from "@fuels-ui/test-utils";
+
+import App from "~/App";
+
+export const checkFirstLoad = async () => {
+  await renderWithRouter(<App />, { route: "/pool/list" });
```
Comment (pedronauck) [inline packages/app/src/pages/WelcomePage/WalletPage.test.tsx:12]: why do you need to be in `/pool/list` route if you're testing Wallet-related stuff?

File: packages/app/src/pages/WelcomePage/WalletPage.test.tsx [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+
+  const createWalletBtn = await screen.findByTestId("createWalletBtn");
+  expect(createWalletBtn).toBeInTheDocument();
+};
+
+export const createWallet = async () => {
```
Comment (pedronauck) [inline packages/app/src/pages/WelcomePage/WalletPage.test.tsx:18]: I think you're making wrong tests here, for `WalletPage` you need to test things related to the wallet widget in the header... if you want to test creating wallet, add funds and all welcome steps, you need to create these tests in the `Welcome` system

File: packages/app/src/pages/PoolPage/PoolPage.test.tsx [hunk 12]
```diff
@@ -1,5 +1,61 @@
...
+    );
+    expect(addLiquiditySubmitBtn).toBeInTheDocument();
+  });
+};
+
 describe("PoolPage", () => {
-  it("pass tests", () => {
-    expect(true);
+  it("should validate pool flow", async () => {
+    await createWallet();
```
Comment (pedronauck) [inline packages/app/src/pages/PoolPage/PoolPage.test.tsx:40]: I really don't think you need to create a wallet and make all steps Welcome-related necessary steps here, you're testing Pools page, everything that isn't related to Pool, you can mock. If you always make all steps on each scenario, our suite will take so much to run.

Like I did in the old tests you removed:

```js
const { privateKey } = Wallet.generate({ provider: FUEL_PROVIDER_URL });
const wallet = new Wallet(privateKey, FUEL_PROVIDER_URL);

jest.mock("../../hooks/useWallet", () => ({
  useWallet: jest.fn(() => wallet),
}));

jest.mock("../../hooks/useContract.ts", () => ({
  useContract: () => ExchangeContractAbi__factory.connect(CONTRACT_ID, wallet),
}));
```

**Review by pedronauck — CHANGES_REQUESTED** [review]

File: packages/app/src/hooks/useWelcomeSteps.tsx (hunk 3 above)
Comment (LuizAsFight) [reply packages/app/src/hooks/useWelcomeSteps.tsx]: we needed because without useMemo we had errors in console saying that machine was different between renders. 

but I removed now to check and the errors are not happening anymore. removing it.

File: packages/app/src/hooks/useFaucet.ts (hunk 4 above)
Comment (LuizAsFight) [reply packages/app/src/hooks/useFaucet.ts]: just for the record, we talked about this and changed other places in the application as well to use `balances.refetch();` instead of `refreshBalances();`

File: packages/app/src/pages/WelcomePage/WalletPage.test.tsx (hunk 10 above)
Comment (LuizAsFight) [reply packages/app/src/pages/WelcomePage/WalletPage.test.tsx]: this test is to check if we access any route, user will be redirected to `CreateWallet` page instead. I'm using `/pool/list`, but could by any other route.

File: packages/app/src/pages/WelcomePage/WalletPage.test.tsx (hunk 11 above)
Comment (LuizAsFight) [reply packages/app/src/pages/WelcomePage/WalletPage.test.tsx]: renamed test to Welcome

File: packages/app/src/pages/PoolPage/PoolPage.test.tsx (hunk 12 above)
Comment (LuizAsFight) [reply packages/app/src/pages/PoolPage/PoolPage.test.tsx]: now we have `App.test.tsx` which has a general end-to-end test of the application (with nothing mocked). This is done to test the whole app, instead of small tests of components. For now it's done until the pool part only, but it's nice to continue the work and test the whole app.

we also have the specific tests like `PoolPage.test.tsx`, `WelcomePage.test.tsx`. which mock data in the way you proposed and quickly test small things in the pages

**Review by pedronauck — APPROVED** [review]

File: .github/workflows/pr.yaml [hunk 13]
```diff
@@ -126,6 +128,6 @@ jobs:
         run: |
           pnpm scripts:setup
 
-      - name: Run test coverage
+      - name: Run test
```
Comment (luizstacio) [inline .github/workflows/pr.yaml:131]: Why we are changing to just test?

File: packages/app/.env.test [hunk 14]
```diff
@@ -1,4 +1,5 @@
 VITE_FUEL_PROVIDER_URL=http://localhost:4001/graphql
 VITE_FUEL_FAUCET_URL=http://localhost:4041/dispense
-VITE_CONTRACT_ID=0x7e35610f31ce8f60b8081b8883db920f6169535020e83904d23e7ae35801d9c4
-VITE_TOKEN_ID=0x5490ccd5425acac521924bd537e4a26b2124427f736a7e56da08c486dc676a25
+VITE_FAUCET_RECAPTCHA_KEY=
+VITE_CONTRACT_ID=0x1491d2949e22d7d28e418f7b4c33bf75e3378ece70be585b960b225a6c1f3182
+VITE_TOKEN_ID=0xf7a2eb8a271a3ea73d6fb8756e272f4a1687d9522acf2ded4454f5d0d5f6a660
```
Comment (luizstacio) [inline packages/app/.env.test:5]: The `swayswap-scripts` will automatically set this envs, I think we can zero this;

```suggestion
VITE_CONTRACT_ID=0x0000000000000000000000000000000000000000000000000000000000000000
VITE_TOKEN_ID=0x0000000000000000000000000000000000000000000000000000000000000000
```

File: packages/app/jest.sequence.js [hunk 15]
```diff
@@ -0,0 +1,24 @@
+const Sequencer = require('@jest/test-sequencer').default;
+
+class CustomSequencer extends Sequencer {
+  /**
+   * Sort test to determine order of execution
```
Comment (luizstacio) [inline packages/app/jest.sequence.js:5]: Can you add more comments about why we need to sequence?

File: .github/workflows/pr.yaml [hunk 16]
```diff
@@ -94,7 +94,9 @@ jobs:
         uses: pnpm/action-setup@v2.1.0
         with:
           version: ${{ env.PNPM_VERSION }}
-          run_install: true
+          run_install: |
+            - recursive: true
+              args: [--no-frozen-lockfile]
```
Comment (luizstacio) [inline .github/workflows/pr.yaml:99]: On the application is a must-have to install using lock files... why do you have to remove it?

File: packages/app/src/systems/Welcome/components/WelcomePage.test.tsx [hunk 17]
```diff
@@ -0,0 +1,20 @@
...
+  it("when first loaded, any route will redirect user to Create Wallet page", async () => {
+    await renderWithRouter(<App />, { route: "/pool/list" });
+    await checkFirstLoad();
+  });
+
+  it("should do create wallet flow", async () => {
+    await renderWithRouter(<App />, {
+      route: "/welcome/create-wallet",
+    });
+    await createWallet();
```
Comment (luizstacio) [inline packages/app/src/systems/Welcome/components/WelcomePage.test.tsx:18]: I think all these functions that abstracts `tests` should start with `expect` or `test`. I normally don't see this type of abstraction and all the code is inside the test.

For example:  `testCreateWallet()` or `expectCreateWalletToBeValid()` this names are more readable and give more context of what  `createWallet` does.

File: packages/config/eslint.js [hunk 18]
```diff
@@ -77,7 +77,7 @@ module.exports = {
   // Disable no-unused-expressions to allow chai 'expect' expressions in testing
   overrides: [
     {
-      files: ['*.test.{ts,tsx}'],
+      files: ['*.test.{ts,tsx}', 'tests.tsx'],
```
Comment (luizstacio) [inline packages/config/eslint.js:80]: maybe you can add the folder `__tests__` with all extensions.

File: packages/test-utils/src/mocks/localstorage.ts [hunk 19]
```diff
@@ -1,9 +1,23 @@
-export function mockLocalStorage(value: string) {
-  Object.defineProperty(window, 'localStorage', {
+export function mockLocalStorage(initialStore: { [key: string]: string }) {
+  let store = initialStore;
```
Comment (luizstacio) [inline packages/test-utils/src/mocks/localstorage.ts:2]: this is ok but could lead to miss interpretation, remember that jest don't share process between tests.

**Review by luizstacio — CHANGES_REQUESTED** [review]

File: package.json [hunk 20]
```diff
@@ -21,8 +21,8 @@
   "scripts": {
     "build": "turbo run build",
     "dev": "turbo run dev --parallel",
-    "ci:test-coverage": "NODE_ENV=test run-s services:run-test contracts test:coverage services:clean-test",
-    "ci:test": "NODE_ENV=test run-s services:run-test contracts test services:clean-test",
+    "ci:test-coverage": "NODE_ENV=test CI=true run-s services:run-test contracts test:coverage services:clean-test",
```
Comment (pedronauck) [inline package.json:24]: you can add this in the action instead of here

**Review by pedronauck — CHANGES_REQUESTED** [review]

File: package.json (hunk 20 above)
Comment (LuizAsFight) [reply package.json:24]: we need to be able to run it locally too, better to keep in scripts

File: .github/workflows/pr.yaml (hunk 13 above)
Comment (LuizAsFight) [reply .github/workflows/pr.yaml:131]: because test coverage is broken, I think we need to fix it and put coverage later

File: packages/app/src/systems/Welcome/components/WelcomePage.test.tsx (hunk 17 above)
Comment (pedronauck) [reply packages/app/src/systems/Welcome/components/WelcomePage.test.tsx:18]: > I think all these functions that abstract tests should start with expect or test. I normally don't see this type of abstraction and all the code is inside the test.

This is the point for me as well

File: packages/app/src/systems/Pool/pages/PoolPage.test.tsx [hunk 21]
```diff
@@ -1,5 +1,137 @@
...
+  return {
+    ...jest.requireActual("../../Core/hooks/useBalances.ts"),
+    useBalances: jest.fn(originalModule.useBalances),
+  };
+});
+
 describe("PoolPage", () => {
-  it("pass tests", () => {
-    expect(true);
+  beforeEach(() => {});
```
Comment (pedronauck) [inline packages/app/src/systems/Pool/pages/PoolPage.test.tsx:59]: why we need this?

**Review by pedronauck — CHANGES_REQUESTED** [review]

File: .github/workflows/pr.yaml (hunk 16 above)
Comment (LuizAsFight) [reply .github/workflows/pr.yaml:99]: we had some problems with dependencies and node version. now with the refactor we might not need it anymore. I'll revert it and check if it works

File: packages/app/src/systems/Pool/pages/PoolPage.test.tsx (hunk 21 above)
Comment (LuizAsFight) [reply packages/app/src/systems/Pool/pages/PoolPage.test.tsx:59]: we don't xD

**Review by luizstacio — APPROVED** [review]

**Review by pedronauck — APPROVED** [review]
